pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,987 @@
+					.ok_or(Error::<T>::LiquidityOverflow)?;
+			}
+
+			T::MultiCurrency::transfer(asset_a, &who, &pool_id, amount_a)?;
+			T::MultiCurrency::transfer(asset_b, &who, &pool_id, amount_b)?;
+
+			Self::deposit_event(Event::LiquidityAdded(pool_id, asset_a, asset_b, amount_a, amount_b));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn remove_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			amount_a: (AssetId, BalanceOf<T>),
+			amount_b: (AssetId, BalanceOf<T>),
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let (asset_a, asset_b) = (amount_a.0, amount_b.0);
+			let (amount_a, amount_b) = (amount_a.1, amount_b.1);
+
+			let asset_pair_id = T::AssetPairPoolId::from_assets(asset_a, asset_b);
+			ensure!(pool_id == asset_pair_id, Error::<T>::InvalidAsset);
+
+			ensure!(
+				!amount_a.is_zero() || !amount_b.is_zero(),
+				Error::<T>::CannotRemoveZeroLiquidity
+			);
+
+			let pool_data = <PoolData<T>>::try_get(&pool_id).map_err(|_| Error::<T>::PoolNotFound)?;
+
+			ensure!(who == pool_data.owner, Error::<T>::NotOwner);
+
+			ensure!(!Self::is_pool_running(&pool_data), Error::<T>::SaleNotEnded);
+
+			if !amount_a.is_zero() {
+				ensure!(
+					T::MultiCurrency::free_balance(asset_a, &pool_id) >= amount_a,
+					Error::<T>::InsufficientAssetBalance
+				);
+
+				T::MultiCurrency::free_balance(asset_a, &who)
+					.checked_add(amount_a)
+					.ok_or(Error::<T>::Overflow)?;
+			}
+
+			if !amount_b.is_zero() {
+				ensure!(
+					T::MultiCurrency::free_balance(asset_b, &pool_id) >= amount_b,
+					Error::<T>::InsufficientAssetBalance
+				);
+
+				T::MultiCurrency::free_balance(asset_b, &who)
+					.checked_add(amount_b)
+					.ok_or(Error::<T>::Overflow)?;
+			}
+
+			T::MultiCurrency::transfer(asset_a, &pool_id, &who, amount_a)?;
+			T::MultiCurrency::transfer(asset_b, &pool_id, &who, amount_b)?;
+
+			Self::deposit_event(Event::LiquidityRemoved(pool_id, asset_a, asset_b, amount_a, amount_b));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::destroy_pool())]
+		#[transactional]
+		pub fn destroy_pool(origin: OriginFor<T>, pool_id: PoolId<T>) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let pool_data = <PoolData<T>>::try_get(&pool_id).map_err(|_| Error::<T>::PoolNotFound)?;
+
+			ensure!(who == pool_data.owner, Error::<T>::NotOwner);
+
+			ensure!(!Self::is_pool_running(&pool_data), Error::<T>::SaleNotEnded);
+
+			let (asset_a, asset_b) = pool_data.assets;
+
+			let amount_a = T::MultiCurrency::free_balance(asset_a, &pool_id);
+			let amount_b = T::MultiCurrency::free_balance(asset_b, &pool_id);
+
+			T::MultiCurrency::transfer(asset_a, &pool_id, &who, amount_a)?;
+			T::MultiCurrency::transfer(asset_b, &pool_id, &who, amount_b)?;
+
+			let deposit = Self::pool_deposit(&pool_id);
+			T::MultiCurrency::unreserve(T::NativeAssetId::get(), &who, deposit);
```

Comment:
I would also check if all deposit has been unreserved. 

You can do so by checking the return value.  And fire an event if there is discrepancy.

Reply:
Didn't know that `unreserve` can "fail". Thanks. But this issue is not relevant anymore. Closing...